src/server.ts

```diff
@@ -13,107 +12,127 @@ declare interface Context {
 }
 
 declare interface QueryTreeArgument {
-  component: any;
-  queries?: any[];
-  context?: Context;
+  rootElement: any;
+  rootContext?: Context;
 }
 
-export function getPropsFromChild(child) {
-  const { props, type } = child;
-  let ownProps = assign({}, props);
-  if (type && type.defaultProps) ownProps = assign({}, type.defaultProps, props);
-  return ownProps;
+declare interface QueryResult {
+  query: Promise<any>;
+  element: any;
+  context: any;
 }
 
-export function getChildFromComponent(component) {
-  // See if this is a class, or stateless function
-  if (component && component.render) return component.render();
-  return component;
-}
-
-let contextStore = {};
-function getQueriesFromTree(
-  { component, context = {}, queries = []}: QueryTreeArgument, fetch: boolean = true
+// Recurse an React Element tree, running visitor on each element.
+// If visitor returns `false`, don't call the element's render function
+//   or recurse into it's child elements
+export function walkTree(
+  element: any,
+  context: any,
+  visitor: (element: any, context: any) => boolean | void
 ) {
-  contextStore = assign({}, contextStore, context);
-  if (!component) return;
-
-  // stateless function
-  if (typeof component === 'function') component = { type: component };
-  const { type, props } = component;
-
-  if (typeof type === 'function') {
-    let ComponentClass = type;
-    let ownProps = getPropsFromChild(component);
-    const Component = new ComponentClass(ownProps, context);
-    try {
-      Component.props = ownProps;
-      Component.context = context;
-      Component.setState = (newState: any) => {
-        Component.state = assign({}, Component.state, newState);
+  // console.log(element)
+  const shouldContinue = visitor(element, context);
+
+  if (shouldContinue === false) {
+    return;
+  }
+
+  const Component = element.type;
+  // a stateless functional component or a class
+  if (typeof Component === 'function') {
+    const props = assign({}, Component.defaultProps, element.props);
+    let childContext = context;
+    let child;
+
+    // Are we are a react class?
+    //   https://github.com/facebook/react/blob/master/src/renderers/shared/stack/reconciler/ReactCompositeComponent.js#L66
+    if (Component.prototype && Component.prototype.isReactComponent) {
+      const instance = new Component(props, context);
+
+      // Override setState to just change the state, not queue up an update.
+      //   (we can't do the default React thing as we aren't mounted "properly"
+      //   however, we don't need to re-render as well only support setState in
+      //   componentWillMount, which happens *before* render).
+      instance.setState = (newState) => {
```

Comment:
Ahhh, I see. That makes sense, I believe. 

I imagine another option might be to run `Component.flushState()` or whatever it's called, just before `Component.componentWillMount()` ? But I don't necessarily see that as more elegant. 

Again, thanks for taking the time to explain!